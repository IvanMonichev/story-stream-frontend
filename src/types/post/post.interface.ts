import { Comment } from '@/types/comment/comment.interface'
import { Id } from '@/types/core/id.type'
import { User } from '@/types/user/user.interface'

export interface Post {
  id: Id
  title: string
  body: string
  likes: { id: Id; user: User }[]
  favoritesCount: 0
  createdAt: Date
  updatedAt: Date
  deletedAt: Date
  user: User
  comments: Comment[]
}
