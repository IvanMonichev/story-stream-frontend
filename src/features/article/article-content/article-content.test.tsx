import { useAuth } from '@/contexts/auth/use-auth.hook'
import { AuthStatus } from '@/enum/core/auth-status.enum'

import ArticleContent from '@/features/article/article-content/article-content'
import { Post } from '@/types/post/post.interface'
import { User } from '@/types/user/user.interface'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { render } from '@testing-library/react'
import React from 'react'
import { afterEach, describe, expect, it, MockedFunction, vi } from 'vitest'

vi.mock('@/contexts/auth/use-auth.hook', () => ({
  __esModule: true,
  useAuth: vi.fn()
}))
// Stub out child components
vi.mock('@/features/article/article-tools/article-tools', () => ({
  __esModule: true,
  default: ({ children }: { children: React.ReactNode }) => (
    <div>{children}</div>
  )
}))
vi.mock('@/shared/ui/button-icon/button-icon', () => ({
  __esModule: true,
  default: ({ children }: { children: React.ReactNode }) => (
    <button>{children}</button>
  )
}))
vi.mock('@/shared/widgets/modal/modal', () => ({
  __esModule: true,
  default: ({
    open,
    children
  }: {
    open: boolean
    children: React.ReactNode
  }) => (open ? <div>{children}</div> : null)
}))
vi.mock('@/features/article/form-article/form-article', () => ({
  __esModule: true,
  default: () => <div />
}))

const mockUseAuth = useAuth as unknown as MockedFunction<typeof useAuth>

// Create a wrapper that provides React Query context
const queryClient = new QueryClient()
const wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => (
  <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
)

describe('<ArticleContent /> snapshots', () => {
  const author: User = { id: 'user1', username: 'john' }
  const otherUser: User = { id: 'user2', username: 'jane' }

  const baseAuthContext = {
    token: null as string | null,
    login: vi.fn() as () => void,
    logout: vi.fn() as () => void
  }

  const sampleArticle: Post = {
    id: 'post1',
    title: 'Test Title',
    body: 'Sample body',
    likes: [],
    favoritesCount: 0,
    createdAt: new Date('2025-05-01T00:00:00Z'),
    updatedAt: new Date('2025-05-02T00:00:00Z'),
    deletedAt: new Date('1970-01-01T00:00:00Z'),
    user: author,
    comments: []
  }

  afterEach(() => {
    vi.clearAllMocks()
  })

  it('renders nothing when no article', () => {
    mockUseAuth.mockReturnValue({
      ...baseAuthContext,
      authStatus: AuthStatus.PENDING,
      user: null
    })
    const { container } = render(<ArticleContent />, { wrapper })
    expect(container).toMatchSnapshot()
  })

  it('renders article for authenticated author', () => {
    mockUseAuth.mockReturnValue({
      ...baseAuthContext,
      authStatus: AuthStatus.AUTHENTICATED,
      user: author
    })
    const { container } = render(<ArticleContent article={sampleArticle} />, {
      wrapper
    })
    expect(container).toMatchSnapshot()
  })

  it('renders article without controls for non-author', () => {
    mockUseAuth.mockReturnValue({
      ...baseAuthContext,
      authStatus: AuthStatus.AUTHENTICATED,
      user: otherUser
    })
    const { container } = render(<ArticleContent article={sampleArticle} />, {
      wrapper
    })
    expect(container).toMatchSnapshot()
  })
})
