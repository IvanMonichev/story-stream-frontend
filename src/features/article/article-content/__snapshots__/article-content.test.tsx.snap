// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ArticleContent /> snapshots > renders article for authenticated author 1`] = `
<div>
  <div
    class="_article_33f2ba"
  >
    <div
      class="_header_33f2ba"
    >
      <div>
        <button>
          Редактировать статью
        </button>
        <button>
          Удалить статью
        </button>
      </div>
      <div
        class="_metaBlock_33f2ba"
      >
        <p
          class="_author_33f2ba"
        >
          John
        </p>
        <div
          class="_date_33f2ba"
        >
          <p
            class="_date_33f2ba"
          >
            01 May, 2025
             год
          </p>
        </div>
      </div>
    </div>
    <div
      class="_body_33f2ba"
    >
      Sample body
    </div>
    <div
      class="_comments_96bb91"
    >
      <h3
        class="_commentsTitle_96bb91"
      >
        Комментарии
      </h3>
      <ul
        class="_commentsList_96bb91"
      />
      <form
        class="_form_8ca615"
        novalidate=""
      >
        <div
          class="_item_4aedc2"
        >
          <label
            class="_label_4aedc2"
            for="content"
          />
          <textarea
            class="_textarea_48b611"
            name="content"
            placeholder="Напишите комментарий"
          />
        </div>
        <button
          class="_button_b92fba"
          type="submit"
        >
          Отправить
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`<ArticleContent /> snapshots > renders article without controls for non-author 1`] = `
<div>
  <div
    class="_article_33f2ba"
  >
    <div
      class="_header_33f2ba"
    >
      <div
        class="_metaBlock_33f2ba"
      >
        <p
          class="_author_33f2ba"
        >
          John
        </p>
        <div
          class="_date_33f2ba"
        >
          <p
            class="_date_33f2ba"
          >
            01 May, 2025
             год
          </p>
        </div>
      </div>
    </div>
    <div
      class="_body_33f2ba"
    >
      Sample body
    </div>
    <div
      class="_comments_96bb91"
    >
      <h3
        class="_commentsTitle_96bb91"
      >
        Комментарии
      </h3>
      <ul
        class="_commentsList_96bb91"
      />
      <form
        class="_form_8ca615"
        novalidate=""
      >
        <div
          class="_item_4aedc2"
        >
          <label
            class="_label_4aedc2"
            for="content"
          />
          <textarea
            class="_textarea_48b611"
            name="content"
            placeholder="Напишите комментарий"
          />
        </div>
        <button
          class="_button_b92fba"
          type="submit"
        >
          Отправить
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`<ArticleContent /> snapshots > renders nothing when no article 1`] = `<div />`;
